### PR TITLE
Feature #2536 : "READ_ENUM_KEYS_USING_INDEX" to match "WRITE_ENUM_KEYS_USING_INDEX"

### PR DESCRIPTION
**Pull request by JooHyukKim**
## Description

- Add enum features to deserialize using enum index
- Matches  `WRITE_ENUM_KEYS_USING_INDEX` symmetrically, (test case included)

**Related Issue** : Need a Feature like "READ_ENUM_KEYS_USING_INDEX" #2536 

Let me know for anything, cheers

### CODE REVIEW

File: src/main/java/com/fasterxml/jackson/databind/deser/std/StdKeyDeserializer.java
Comment (yawkat) [inline src/main/java/com/fasterxml/jackson/databind/deser/std/StdKeyDeserializer.java:374]:
```suggestion
```

File: src/main/java/com/fasterxml/jackson/databind/deser/std/StdKeyDeserializer.java
Comment (yawkat) [inline src/main/java/com/fasterxml/jackson/databind/deser/std/StdKeyDeserializer.java:436]: this form of double-checked locking is not thread-safe, see https://shipilev.net/blog/2014/safe-public-construction/

_byIndexResolver needs to be volatile

File: src/main/java/com/fasterxml/jackson/databind/util/EnumResolver.java
Comment (yawkat) [inline src/main/java/com/fasterxml/jackson/databind/util/EnumResolver.java:149]:
```suggestion
    /**
```

File: src/main/java/com/fasterxml/jackson/databind/util/EnumResolver.java
Comment (yawkat) [inline src/main/java/com/fasterxml/jackson/databind/util/EnumResolver.java:130]: indent looks wrong at least on github

File: src/test/java/com/fasterxml/jackson/databind/deser/enums/EnumDeserializationTest.java
Comment (pjfanning) [inline src/test/java/com/fasterxml/jackson/databind/deser/enums/EnumDeserializationTest.java:658]: the 3 space indents do not match up with the 4 space indents used - it seems in all of these files

File: src/main/java/com/fasterxml/jackson/databind/util/EnumResolver.java
Comment (pjfanning) [reply src/main/java/com/fasterxml/jackson/databind/util/EnumResolver.java:130]: user seems to be using 3 space indents instead of 4 space indents that are used in existing code

Comment (yawkat) [reply src/main/java/com/fasterxml/jackson/databind/util/EnumResolver.java:130]: i don't have to visit my optometrist then :sweat_smile:

Comment (JooHyukKim) [reply src/main/java/com/fasterxml/jackson/databind/util/EnumResolver.java:130]: 😂 My bad for the visual trickery

Comment (JooHyukKim) [reply src/main/java/com/fasterxml/jackson/databind/util/EnumResolver.java:130]: @pjfanning @yawkat
Just to make sure, is it 4 space indents? I think I references the wrong classes for indents

File: src/test/java/com/fasterxml/jackson/databind/deser/enums/EnumDeserializationTest.java
Comment (JooHyukKim) [reply src/test/java/com/fasterxml/jackson/databind/deser/enums/EnumDeserializationTest.java]: done~ my apologies for the confusion  😆

File: src/main/java/com/fasterxml/jackson/databind/deser/std/StdKeyDeserializer.java
Comment (pjfanning) [inline src/main/java/com/fasterxml/jackson/databind/deser/std/StdKeyDeserializer.java:435]: I think it would be better to not have the `res` local variable and to refer to `_byIndexResolver ` directly throughout this method

File: src/main/java/com/fasterxml/jackson/databind/deser/std/StdKeyDeserializer.java
Comment (pjfanning) [inline src/main/java/com/fasterxml/jackson/databind/deser/std/StdKeyDeserializer.java:439]: space before `{`

File: src/main/java/com/fasterxml/jackson/databind/deser/std/StdKeyDeserializer.java
Comment (pjfanning) [inline src/main/java/com/fasterxml/jackson/databind/deser/std/StdKeyDeserializer.java:443]: should be _byIndexResolver

File: src/main/java/com/fasterxml/jackson/databind/deser/std/StdKeyDeserializer.java
Comment (pjfanning) [inline src/main/java/com/fasterxml/jackson/databind/deser/std/StdKeyDeserializer.java:436]: could you synchronize this like you were asked by @yawkat in the other PR?

**Review by pjfanning — CHANGES_REQUESTED** [review]

**Review by pjfanning — APPROVED** [review]
lgtm but will need @cowtowncoder's approval to merge

**Comment by JooHyukKim** [comment]
@cowtowncoder do you have any other opinion on this PR?

**Comment by cowtowncoder** [comment]
> @cowtowncoder do you have any other opinion on this PR?

Unfortunately I haven't had time to look into it yet -- there's a big backlog of items. But I hope to have a look soon. Thank you for contributing this!

File: src/main/java/com/fasterxml/jackson/databind/DeserializationFeature.java
Comment (cowtowncoder) [inline src/main/java/com/fasterxml/jackson/databind/DeserializationFeature.java:415]: add `@since 2.15`

**Comment by cowtowncoder** [comment]
Ok, one quick note: I do not want to add a new `DeserializationFeature`, as those should ideally be more general, not datatype-specific: I know a few already exists but I hope they eventually get moved out (for 3.0).

Instead I would like it added using `EnumFeature.READ_ENUM_KEYS_USING_INDEX`; similar to how new options have been added using `JsonNodeFeature`.
That should be about as easy to use and offers more extensibility (number of `DeserializationFeature` entries is limited to 31, f.ex).

File: src/main/java/com/fasterxml/jackson/databind/util/EnumResolver.java
Comment (cowtowncoder) [inline src/main/java/com/fasterxml/jackson/databind/util/EnumResolver.java:129]: Need to add Javadoc that includes `@since 2.15` to indicate when method was added.

**Comment by cowtowncoder** [comment]
Aside from `DeserializationFeature` -> `EnumFeature`, and need for minor addition of Javadocs (to indicate version public things were added in), looks pretty good! Good job.
Will be happy to merge after changes.

**Comment by JooHyukKim** [comment]
> Ok, one quick note: I do not want to add a new `DeserializationFeature`, as those should ideally be more general, not datatype-specific: I know a few already exists but I hope they eventually get moved out (for 3.0).

I see, these feature are to be moved out. So do we not need this **symmetry** with `SerializationFeature.WRITE_ENUMS_USING_INDEX` also mentioned in #2536 ? just curious

> Instead I would like it added using `EnumFeature.READ_ENUM_KEYS_USING_INDEX`; similar to how new options have been added using `JsonNodeFeature`. That should be about as easy to use and offers more extensibility (number of `DeserializationFeature` entries is limited to 31, f.ex).

Noted, I will work on it.

**Review by cowtowncoder — APPROVED** [review]

**Comment by cowtowncoder** [comment]
Thank you @JooHyukKim for contributing this -- I finally had time to merge it, and it will go in 2.15.0 release.

**Pull request merged**
